### PR TITLE
Convert `gitserver.DefaultClient` to an interface

### PR DESCRIPTION
**Pull request by camdencheek**
This creates a `gitserver.Client` interface, and converts `gitserver.DefaultClient` to be that type. The motivation here is that I really want to be able to swap out `DefaultGitserver` with a mocked version in search tests, but that's currently not possible. Ideally, we'd go as far as to pass in a `gitserver.Client` object to the search code, but we're not there yet. 

To do this, I needed to get rid of places where the publically-exposed `Addrs` field is mutated and make it private. This meant making a new method `NewTestClient` that lets you pass in a hard-coded set of `addrs`. 

Ideally, with a real interface type, we will be able to pass that type around through code that needs tested so we can just pass in a generated mock implementation instead. 

@ryanslade Added you as a reviewer because IIRC, you had/have been working on some hexagonal architecture stuff for gitserver

### CODE REVIEW

File: enterprise/cmd/frontend/internal/codemonitors/resolvers/resolvers_test.go [hunk 1]
```diff
@@ -673,7 +673,7 @@ func TestEditCodeMonitor(t *testing.T) {
 			CreatedBy: apitest.UserOrg{
 				Name: user1.Username,
 			},
-			CreatedAt: marshalDateTime(t, r.db.CodeMonitors().Now()),
```
Comment (camdencheek) [inline enterprise/cmd/frontend/internal/codemonitors/resolvers/resolvers_test.go:676]: unrelated, but fixing a flaky test

**Comment by keegancsmith** [comment]
Just my usual band wagon, but why not just an interface for what search needs instead of exporting this monster interface? I've seen this done by a few code-intel things and batch changes I believe. example here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@91ebac286f1b6cd034ede54d45beef10016c6f46/-/blob/cmd/symbols/internal/gitserver/client.go?L19-27&subtree=true

Another fun thing is instead of using mocks use a fake client which speaks to a git repo on disk. I think we do that in some places already, if not a bit clumsy these days. That way you test actual contracts in git instead of a bunch of fake data setup in a mock. IE smaller and more reliable tests.

But obviously won't block this PR, just thought I'd spout an alternative as a drive-by comment.

**Comment by camdencheek** [comment]
> but why not just an interface for what search needs instead of exporting this monster interface?

In my view, the two approaches are complementary. Ideally, yes: search should scope its dependency down to a smaller interface. However, having an interface type for the whole client means that we can generate a mock for the whole client, and that generated mock can be used for any subset of the interface it is generated for. That way we don't have to generate 20 different mock clients for the 20 different subsets of the interface needed throughout the codebase.

However, the more practical reason right now is that we use `gitserver.DefaultClient` [all over the place](https://sourcegraph.com/search?q=context:global+repo:github.com/sourcegraph/sourcegraph%24+gitserver.DefaultClient&patternType=literal) in the codebase. I don't have the bandwidth to inject a smaller client everywhere at this point, so instead I'm taking a first step to making it testable: making the global replaceable.

**Comment by ryanslade** [comment]
FYI @sourcegraph/repo-management

**Comment by ryanslade** [comment]
I have similar concerns regarding exposing the huge interface as Keegan but agree that for practical reasons right now it's the easies way forward.

**Review by ryanslade — APPROVED** [review]

**Comment by camdencheek** [comment]
I'd be more than happy to see someone take this in a different direction, but given this is ready to go and will unblock me from testing a bunch of stuff that depends on gitserver, I'll to merge this as-is. Certainly not tied to the gigantic interface though 😄

**Pull request merged**